server-go/main.go

```diff
@@ -0,0 +1,234 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"cmp"
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"strings"
+
+	"github.com/google/generative-ai-go/genai"
+	"github.com/rs/cors"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+)
+
+const modelName = "gemini-1.5-flash"
+const defaultPort = "9000"
+
+// Server state holding the context of the Gemini client and the generative model.
+type genaiServer struct {
+	ctx   context.Context
+	model *genai.GenerativeModel
+}
+
+func main() {
+	ctx := context.Background()
+
+	// Access your API key as an environment variable to create a client.
+	apiKey := os.Getenv("GOOGLE_API_KEY")
+	client, err := genai.NewClient(ctx, option.WithAPIKey(apiKey))
+	if err != nil {
+		log.Fatalf("could not create Gemini client %v", err)
+	}
+	defer client.Close()
+
+	model := client.GenerativeModel(modelName)
+
+	server := &genaiServer{
+		ctx:   ctx,
+		model: model,
+	}
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("POST /chat", server.chatHandler)
+	mux.HandleFunc("POST /stream", server.streamingChatHandler)
+
+	// Add CORS middleware handler.
+	c := cors.New(cors.Options{
+		AllowedOrigins: []string{"*"},
+		AllowedHeaders: []string{"Access-Control-Allow-Origin", "Content-Type"},
+	})
+	handler := c.Handler(mux)
+
+	// Access preferred port the server must listen to as an environment variable if provided.
+	port := cmp.Or(os.Getenv("PORT"), defaultPort)
+	addr := "localhost:" + port
+	log.Println("Listening on ", addr)
+	log.Fatal(http.ListenAndServe(addr, handler))
+}
+
+// part is a piece of model content. It can hold only text pieces. Each item in the `Parts` of a
+// JSON-encoded content in a history array must comply to it.
```

Comment:
I don't quite understand this, but it seems important: "Each item in the `Parts` of a
// JSON-encoded content in a history array must comply to it." Can you maybe clarify or rephrase?